Jack Networking

jhar23 opened the pull request:
Created A way for data to be sent in different time intervals to base station

rgw3d committed:
Merge pull request #40 from IlliniHyperloopComputing/pod3

Merge pod3 into master

jhar23 committed:
Added val1 and val2, which are reconstructed vals sent from the pod

jhar23 committed:
Showing how parsing bytes on python side could go

jhar23 committed:
Created a way to push different values at different times in TCPManager. Made tcpserver.py print out the bytes of bytearray

jhar23 committed:
Merge remote-tracking branch 'origin' into Jack_Networking

rgw3d commented on CentralComputing/TCPManager.cpp:
```diff
@@ -55,14 +55,30 @@ int TCPManager::read_command(Network_Command * buffer) {
 }
 
-int TCPManager::write_data() {
+int TCPManager::write_data(vector<int64_t>& times) {
```
Make this a `const` reference!

rgw3d commented on CentralComputing/TCPManager.cpp:
```diff
@@ -55,14 +55,30 @@ int TCPManager::read_command(Network_Command * buffer) {
-int TCPManager::write_data() {
+int TCPManager::write_data(vector<int64_t>& times) {
+  vector<int32_t> vals;
+  if(Utils::microseconds() - times[0] > 1000000){  //  This is the first time threshold
```
I suggest calling `Utils::microseconds()` just once, and then using that value repeatedly

rgw3d commented on CentralComputing/TCPManager.h:
```diff
@@ -85,9 +85,10 @@ int read_command(Network_Command * buffer);
  * @return bytes written or -1 if failed
  **/
-int write_data();
+int write_data(std::vector<int64_t>& times);
```
Maybe rename to `last_sent_time` instead of `times`, or something like that. Makes it a bit easier to understand.

rgw3d commented on CentralComputing/TCPManager.cpp:
```diff
@@ -55,14 +55,30 @@ int TCPManager::read_command(Network_Command * buffer) {
+    vals.push_back(ACCELERATION);
+    times[0] = Utils::microseconds();
+  }
+  if(Utils::microseconds() - times[1] > 3000000){  //  This is the second time threshold 
```
the `3000000` delay can be made configurable. That is, you can load in a value from the configuration file instead of having a "magic" number in the code. 

Take a look at `pod.cpp` to see how it loads in data from the configuration file. What you will want to do is to create another array (defined in the TCPManager namespace) that is initialized with the configurations either at the start of `write_loop()` or at the start of `tcp_loop()` (probably tcp_loop()).

rgw3d replied:
You will also need to define these delays within the configuration file. Name that whatever you want! 

This makes it really easy for us down the road to simply change the delay time without having to touch the code.

rgw3d commented on CentralComputing/TCPManager.cpp:
```diff
@@ -55,14 +55,30 @@ int TCPManager::read_command(Network_Command * buffer) {
+    vals.push_back(TEMPERATURE);
+    times[1] = Utils::microseconds();
+  }
+  if(Utils::microseconds() - times[2] > 6000000){  //  This is the third time threshold 
```
Ideally, there is some way to turn these `if`s into something that you can put into a for loop, since all of the conditions in the `if` can be parameterized to an index. 

I'm not sure exactly how you would get the `vals.push_back` to be parameratized by index, and it might not be worth the effort. 

For now, this is a very convienent and easy way of doing things.

rgw3d reviewed (changes requested):
Looking good!

jhar23 committed:
Merge branch 'pod4' of https://github.com/IlliniHyperloopComputing/Pod into Jack_Networking

jhar23 committed:
Fixing merging problems

jhar23 committed:
Made first big changes from Pull Request revisions.

First thing done was calling Utils::microseconds() once and
storing it in a variable. Next, was getting rid of the
parameter passed into write_data, and instead use
a static variable to hold the times. Finally
I added the stagger times into defaultConfig.txt
and will update them in the TCPManager in the
next commit!

jhar23 committed:
Finished all revisions except for last revision

The revisions of the commit are: implementing the
defaultconfig.txt values into write_data, changing the
name of times to last_sent_times for checking when to
send data. The first revision was done by adding a
ConfiguratorManager check and write to the array
"stagger_times" inside the beginning of tcp_loop.

rgw3d committed:
Small change to TCP configuration

rgw3d committed:
Fixed the declaration of stagger_times

rgw3d committed:
Last_sent_times[] is not a local variable anymore.

rgw3d committed:
Moving TCPwrite timeouts to Pod.cpp

jhar23 committed:
Moving timeout staggers to Pod.cpp and making Safequeues for each struct

jhar23 committed:
Merge branch 'Jack_Networking' of https://github.com/IlliniHyperloopComputing/Pod into Jack_Networking

rgw3d committed:
Adding objects to queue

jhar23 committed:
Changed write_data to use Utils::write_all_to_socket

jhar23 committed:
Merge branch 'Jack_Networking' of https://github.com/IlliniHyperloopComputing/Pod into Jack_Networking

rgw3d committed:
Adding a concrete type to the E_States define

rgw3d committed:
Removing the stagger_timeouts from Pod.cpp. We did just put them in here, but changing it back

rgw3d committed:
Revamping how to send data. Moving away from Queues to a single variable. Memory is now more fixed, but still not the ideal static buffer that I eventually want.

rgw3d committed:
Adding a configurable write_loop timeout for the write_loop()

rgw3d committed:
Cleaning up and linting

rgw3d committed:
Changing Pod's internal storage of Unified state to a static buffer.

rgw3d committed:
Reflecting changes of Pod's Unified state here

rgw3d committed:
reflecting the changes in UnifiedState and MotionModel's signature here

rgw3d committed:
Creating local buffers to store all information

rgw3d committed:
removed an unecessary helper function

jhar23 committed:
Merge branch 'master' of https://github.com/IlliniHyperloopComputing/Pod into Jack_Networking

jhar23 committed:
Merge branch 'Jack_Networking' of https://github.com/IlliniHyperloopComputing/Pod into Jack_Networking

jhar23 committed:
Allows the TCP to reconnect in the case that it disconnects

rgw3d committed:
Adding in TCP errors on disconnect

rgw3d committed:
Making existing tests work with new TCP errors

rgw3d committed:
Linting

rgw3d committed:
Silencing UDP printouts

rgw3d reviewed (approved):
Good to go!
